debugger_implementation.tex

```diff
@@ -606,7 +606,19 @@ \subsection{Single Step} \label{nativestep}
 \FcsrIcountM=0 before returning control to the lower user program with an
 {\tt mret} instruction.
 
-On tiny systems which only support M-Mode single step is doable, but tricky to
+Stepping code running in the same privilege mode as the debugger is more
+complicated, depending on what other debug features are implemented.
+
+If hardware implements \FcsrTcontrolMpte and \FcsrTcontrolMte, then stepping
+through non-trap code which doesn't allow for nested interrupts is also
+straightforward.
+
+If hardware automatically prevents \FcsrMcontrolSixAction=0 triggers from
+matching when entering a trap handler as described in
+Section~\ref{sec:nativetrigger}, then a carefully written trap handler can
+ensure that interrupts are enabled whenever the icount trigger is enbled.
```

Comment:
```suggestion
ensure that interrupts are enabled whenever the icount trigger is enabled.
```

I wouldn't say that the handler should make sure that interrupts are enabled whenever the icount trigger is enabled.  It should make sure that interrupts are enabled whenever you want to match.  Or to count instructions.  Or whenever the trap handler wants icount to be active.  Or "armed" (though we don't use that word anywhere else so defining that opens a whole new challenge).

The word "enabled" is just too ambiguous for me.  My initial interpretation is just m=1.

Reply:
Updated. A little more double-negative than I prefer, but I'm not sure how to do it otherwise.

Reply:
I see what you mean.  But I think it gets the point across and is unambiguous.